apps/functions-subscription/src/use-cases/insert-subscription.ts

```diff
@@ -0,0 +1,69 @@
+import { pipe } from 'fp-ts/lib/function';
+import * as O from 'fp-ts/Option';
+import * as TE from 'fp-ts/TaskEither';
+import * as RTE from 'fp-ts/ReaderTaskEither';
+import { Capabilities } from '../domain/capabilities';
+import {
+  SubscriptionId,
+  TrialId,
+  UserId,
+  makeSubscriptionId,
+} from '../domain/subscription';
+import { SubscriptionAlreadyExists, SubscriptionStoreError } from './errors';
+import { nowDate } from '../domain/clock';
+
+// Maps all the requirements for this use-case
+type Env = Pick<
+  Capabilities,
+  | 'subscriptionRequestWriter'
+  | 'subscriptionReader'
+  | 'subscriptionWriter'
+  | 'hashFn'
+  | 'clock'
+>;
+
+const handleSubscriptionAlreadyExists =
+  (id: SubscriptionId) => (env: Pick<Capabilities, 'subscriptionReader'>) =>
+    pipe(
+      env.subscriptionReader.get(id),
```

Comment:
What if reader returns an error? this function returns a TaskEither<Error | SubscriptionAlreadyExists, SubscriptionId>. Aside of this, are we going to disjoint error types in the caller code, in order to return mapped http codes?

Reply:
Yes that's right. If needed the caller can maps the error to a defined responses